core/types/calls.cc

```diff
@@ -1363,6 +1363,31 @@ class Magic_buildArray : public IntrinsicMethod {
     }
 } Magic_buildArray;
 
+class Magic_buildRange : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        ENFORCE(!args.args.empty(), "Magic_buildRange called without argument");
+
+        auto rangeElemType = Types::dropLiteral(args.args[0]->type);
+        if (!rangeElemType->isNilClass()) {
+            rangeElemType = Types::dropNil(gs, rangeElemType);
+        }
+
+        if (args.args.size() > 1) {
+            auto other = Types::dropLiteral(args.args[1]->type);
+            if (rangeElemType->isNilClass() && other->isNilClass()) {
+                rangeElemType = Types::untypedUntracked();
+            } else if (rangeElemType->isNilClass()) {
+                rangeElemType = Types::dropNil(gs, other);
+            } else if (other->isUntyped() || (!other->isNilClass() && !Types::equiv(gs, other, rangeElemType))) {
+                rangeElemType = Types::untypedUntracked();
+            }
+        }
+
+        res.returnType = Types::rangeOf(gs, rangeElemType);
+    }
```

Comment:
```suggestion
    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
        ENFORCE(args.args.size() == 2, "Magic_buildRange called without argument");

        auto rangeElemType = Types::dropLiteral(args.args[0]->type);
        auto firstArgIsNil = Types::isSubType(gs, rangeElemType, Types::nilClass());
        if (!firstArgIsNil) {
            rangeElemType = Types::dropNil(gs, rangeElemType);
        }

        auto other = Types::dropLiteral(args.args[1]->type);
        auto secondArgIsNil = Types::isSubType(gs, other, Types::nilClass());
        if (firstArgIsNil) {
            if (secondArgIsNil) {
                rangeElemType = Types::untypedUntracked();
            } else {
                rangeElemType = Types::dropNil(gs, other);
            }
        } else if (other->isUntyped() || (!secondIsNil && !Types::equiv(gs, other, rangeElemType))) {
            rangeElemType = Types::untypedUntracked();
        }

        res.returnType = Types::rangeOf(gs, rangeElemType);
    }
```

* You can enforce that the buildRange intrinsic always has exactly two arguments because the only calls to it are constructed directly in `Desugar.cc` with two arguments
* Knowing that it has two arguments means that the conditional on the argument count at line 1376 will always be true
* You can check to see if the type is a subtype of `NilClass` instead of implementing a custom `isNilClass` function
* What do you think about raising an error in the final else-if case? I think that this would happen when the two bounds of the range are not `nil` or `T.untyped`, which would indicate that the bounds are ill-formed.